pkg/kube/krt/index_test.go

```diff
@@ -0,0 +1,148 @@
+	kpc := kclient.New[*corev1.Pod](c)
+	pc := clienttest.Wrap(t, kpc)
+	pods := krt.WrapClient[*corev1.Pod](kpc)
+	stop := test.NewStop(t)
+	c.RunAndWait(stop)
+	SimplePods := SimplePodCollection(pods)
+	tt := assert.NewTracker[string](t)
+	IPIndex := krt.NewIndex[SimplePod, string](SimplePods, func(o SimplePod) []string {
+		return []string{o.IP}
+	})
+	fetchSorted := func(ip string) []SimplePod {
+		return slices.SortBy(IPIndex.Lookup(ip), func(t SimplePod) string {
+			return t.ResourceName()
+		})
+	}
+
+	SimplePods.Register(TrackerHandler[SimplePod](tt))
+
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "name",
+			Namespace: "namespace",
+		},
+		Status: corev1.PodStatus{PodIP: "1.2.3.4"},
+	}
+	pc.CreateOrUpdateStatus(pod)
+	tt.WaitUnordered("add/namespace/name")
+	assert.Equal(t, fetchSorted("1.2.3.4"), []SimplePod{{NewNamed(pod), Labeled{}, "1.2.3.4"}})
+
+	pod.Status.PodIP = "1.2.3.5"
+	pc.UpdateStatus(pod)
+	tt.WaitUnordered("update/namespace/name")
+	assert.Equal(t, fetchSorted("1.2.3.4"), []SimplePod{})
+	assert.Equal(t, fetchSorted("1.2.3.5"), []SimplePod{{NewNamed(pod), Labeled{}, "1.2.3.5"}})
+
+	pod2 := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "name2",
+			Namespace: "namespace",
+		},
+		Status: corev1.PodStatus{PodIP: "1.2.3.5"},
+	}
+	pc.CreateOrUpdateStatus(pod2)
+	tt.WaitUnordered("add/namespace/name2")
+	assert.Equal(t, fetchSorted("1.2.3.5"), []SimplePod{
+		{NewNamed(pod), Labeled{}, "1.2.3.5"},
+		{NewNamed(pod2), Labeled{}, "1.2.3.5"},
+	})
+
+	pc.Delete(pod.Name, pod.Namespace)
+	pc.Delete(pod2.Name, pod2.Namespace)
+	tt.WaitUnordered("delete/namespace/name", "delete/namespace/name2")
+	assert.Equal(t, fetchSorted("1.2.3.4"), []SimplePod{})
+}
+
+func TestIndexCollection(t *testing.T) {
+	c := kube.NewFakeClient()
+	kpc := kclient.New[*corev1.Pod](c)
+	pc := clienttest.Wrap(t, kpc)
+	pods := krt.WrapClient[*corev1.Pod](kpc)
+	stop := test.NewStop(t)
+	c.RunAndWait(stop)
+	SimplePods := SimplePodCollection(pods)
+	tt := assert.NewTracker[string](t)
+	IPIndex := krt.NewIndex[SimplePod, string](SimplePods, func(o SimplePod) []string {
+		return []string{o.IP}
+	})
+	Collection := krt.NewSingleton[string](func(ctx krt.HandlerContext) *string {
+		pods := krt.Fetch(ctx, SimplePods, krt.FilterIndex(IPIndex, "1.2.3.5"))
+		names := slices.Sort(slices.Map(pods, SimplePod.ResourceName))
+		return ptr.Of(strings.Join(names, ","))
+	})
+	Collection.AsCollection().Synced().WaitUntilSynced(stop)
+	fetchSorted := func(ip string) []SimplePod {
+		return slices.SortBy(IPIndex.Lookup(ip), func(t SimplePod) string {
+			return t.ResourceName()
+		})
+	}
+
+	SimplePods.Register(TrackerHandler[SimplePod](tt))
+
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "name",
+			Namespace: "namespace",
+		},
+		Status: corev1.PodStatus{PodIP: "1.2.3.4"},
+	}
+	pc.CreateOrUpdateStatus(pod)
+	tt.WaitUnordered("add/namespace/name")
+	assert.Equal(t, Collection.Get(), ptr.Of(""))
```

Comment:
Why is this one empty?

Reply:
The index is for `1.2.3.5` so this doesn't match